File: web/src/lib/sensitiveDisplay.ts
/** Redact AWS access key IDs in user-facing copy. Full IDs remain in API evidence for remediation. */

const AWS_ACCESS_KEY_RE = /\b((?:AKIA|ASIA|AROA)[A-Z0-9]{16})\b/g;

export function maskAccessKeyId(keyId: string | null | undefined): string {
  const key = (keyId ?? "").trim();
  if (!key) return key;
  if (key.length <= 8) return "••••";
  return `${key.slice(0, 4)}••••${key.slice(-4)}`;
}

export function maskSensitiveText(text: string | null | undefined): string {
  if (!text) return text ?? "";
  return text.replace(AWS_ACCESS_KEY_RE, (_, key: string) => maskAccessKeyId(key));
}

File: api/app/core/sensitive_display.py
"""Redact secrets in user-facing strings (access keys, etc.)."""
from __future__ import annotations

import re

_AWS_ACCESS_KEY_RE = re.compile(r"\b((?:AKIA|ASIA|AROA)[A-Z0-9]{16})\b")


def mask_access_key_id(key_id: str | None) -> str:
    key_id = (key_id or "").strip()
    if not key_id:
        return key_id
    if len(key_id) <= 8:
        return "••••"
    return f"{key_id[:4]}••••{key_id[-4:]}"


def mask_sensitive_text(text: str | None) -> str:
    if not text:
        return text or ""

    def _sub(match: re.Match[str]) -> str:
        return mask_access_key_id(match.group(1))

    return _AWS_ACCESS_KEY_RE.sub(_sub, text)
